compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -403,12 +406,35 @@ public void visit(BLangResourceFunction funcNode, AnalyzerData data) {
             BLangType restParamType = funcNode.resourcePathType.restParamType;
             symResolver.resolveTypeNode(restParamType, data.env);
             if (!types.isAssignable(restParamType.getBType(), symTable.pathParamAllowedType)) {
-                dlog.error(restParamType.getPosition(), DiagnosticErrorCode.UNSUPPORTED_REST_PATH_PARAM_TYPE, 
+                dlog.error(restParamType.getPosition(), DiagnosticErrorCode.UNSUPPORTED_REST_PATH_PARAM_TYPE,
                         restParamType.getBType());
             }
         }
     }
 
+    private void checkForClientObjectTypeOrFunctionType(BType type, BType referredType, Location pos) {
```

Comment:
Can't we refactor the `checkForClientObjectTypeOrFunctionType` function as
```java
private void checkForClientObjectTypeOrFunctionType(BType type) {
        BType referredType = Types.getReferredType(type);
        Location pos = type.tsymbol.pos;
        if (referredType.tsymbol != null && Symbols.isFlagOn(referredType.tsymbol.flags, Flags.CLIENT)) {
            dlog.error(pos, DiagnosticErrorCode.INVALID_RESOURCE_METHOD_RETURN_TYPE, referredType);
        }
        switch (referredType.tag) {
            case TypeTags.INVOKABLE:
                dlog.error(pos, DiagnosticErrorCode.INVALID_RESOURCE_METHOD_RETURN_TYPE, referredType);
                break;
            case TypeTags.UNION:
                for (BType memberType: ((BUnionType) referredType).getMemberTypes()) {
                    checkForClientObjectTypeOrFunctionType(memberType);
                }
                break;
            case TypeTags.INTERSECTION:
                for (BType memberType: ((BIntersectionType) referredType).getConstituentTypes()) {
                    checkForClientObjectTypeOrFunctionType(memberType);
                }
        }
    }
```

Reply:
If we implement like this, we cannot get the high level type name for the error message.

Reply:
With this implementation we cannot get the correct type name for the error message.
ex:- 
```
type CustomType0 float|int[]|CustomType3|int|string;
type CustomType1 float|int[]|CustomType3|int|string;
type CustomType3 client object{};
type CustomType4 CustomType3 | CustomType1;

client object{} & readonly successClient = client object {

};

public function testErrorInResourceMethodReturningObjectType() {
    var successClientObject = client object {
        resource function get clientReturnPath() returns CustomType4 {
            return successClient;
        }
    };
}
```

In this code, error message will not show the type name `CustomType4`
It will say something like, `return type of the resource function does not allow 'client object' type: 'object {}'`

Reply:
Can't we keep the original type as a parameter of this function? We need that only for the error message. Right?

Reply:
Yes.
` private void checkForClientObjectTypeOrFunctionType(BType type, BType referredType, Location pos)`
type parameter is for the original type

Reply:
We cannot keep original type as a local variable, because recursive calls can log the error.

Reply:
The structure of the error message got changed. So, no need the original type name for the error message. Therefore +1 with this suggestion. Updated 2725ba33d9230d4c98b3b79818080c3ba31e21b3